### DISCUSSION TITLE
How do I deploy an existing Dagster project scaffolded in a dbt project to Dagster Cloud

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#19071] [Q&A] [answered]
Hey all. I am having some trouble deploying a project to Dagster Cloud. My main goal is to use Dagster to schedule my dbt transformations, but I also have some custom data ingestion functions. I have a working project that I am successfully running locally on my computer, as well as on a kubernetes cluster (the latter I want to move away from, hence Dagster cloud).

But I can’t deploy it to Dagster Cloud, because I can’t seem to get the configuration to work. Unfortunately, the <https://docs.dagster.io/integrations/dbt|documentation> only explains how to either 1) run this locally, or 2) set this up from scratch using an existing dbt project without any Dagster project. I’ll post more details in the thread.

I started with an existing dbt project, after which I used the Dagster documentation to scaffold a Dagster project inside the dbt repo (https://docs.dagster.io/integrations/dbt/using-dbt-with-dagster). This means my project structure looks something like this:

```
dbt_repo 
--analyses
--dagster_analytics
	--dagster_analytics
		--assets
		--constants.py
		--definitions.py
		--schedules.py
	--pyproject.toml
	--setup.py
--macros
--models
--seeds
--snapshots
--target
--tests
--dagster_cloud.yaml
--dbt_project.yml
--profiles.yml
--pyproject.toml
--README.md
```

My `dagster_cloud` file looks like this. I am using this config, because running the following command in the `dbt_repo` root works fine: `dagster dev -m dagster_analytics.definitions -d ./dagster_analytics`.

```yaml
locations:
  - location_name: dagster_analytics # the name of the code location
    code_source:
      module_name: "dagster_analytics.definitions"
    build: 
      directory: ./dagster_analytics
```

However, the CI/CD build is not successful. I am getting a sensible error: `venvs/9d31073e2525/lib/python3.8/site-packages/working_directory does not contain a dbt_project.yml file. Please specify a valid path to a dbt project.`

I am assuming that, since I point to `./dagster_analytics` as a build dir, it’s simply not including anything from the root in the docker image build, so it can’t find the dbt files. How do I make sure that the root folder is included in the CI/CD? And how can I debug this?
I tried setting the working_directory instead of the build directory, like this:

```yaml
locations:
  - location_name: dagster_analytics # the name of the code location
    code_source:
      module_name: "dagster_analytics.definitions"
    working_directory: ./dagster_analytics 
```

But then it cannot find setup.py: `Error: Could not find a `setup.py` or `requirements.txt` in the target directory. You must specify your required Python dependencies (including the `dagster-cloud` package) along with your source files to deploy to Dagster Cloud.`

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C04CW71AGBW/p1704732233543389?thread_ts=1704732233.543389&cid=C04CW71AGBW).*

**Comment by rexledesma** [comment, marked as answer]
I assume you're using Dagster Cloud Serverless. With Dagster Cloud Serverless, we have an onboarding step that automatically creates a Dagster project in an existing dbt project repository.

<img width="1134" alt="Screenshot 2024-02-14 at 2 49 56 PM" src="https://github.com/dagster-io/dagster/assets/16431325/93036366-32c0-47ed-8e2e-dde468475be5">


This scaffolds a project using `dagster-dbt project scaffold` and sets up the necessary configuration for building the project and setting up CI/CD.

In your case, you have already set up the project using `dagster-dbt project scaffold`. What remains is to add the configuration for building the project properly.

There are two steps that need to be done:

1. Configure your Dagster Cloud Serverless package to include your dbt project as part of the package.
2. Copy your dbt project to a subdirectory of your Dagster project.
3. Include the environment variables for your `profiles.yml` to be hosted in your CI/CD system to build your manifest, as well as [Dagster Cloud](https://docs.dagster.io/dagster-cloud/managing-deployments/environment-variables-and-secrets) to execute your dbt project.

---

For step (1), you can follow [this guidance](https://docs.dagster.io/dagster-cloud/deployment/serverless#customizing-the-runtime-environment) for including your dbt project in your package build with PEX. You need to modify your `setup.py` to include this information:

```diff
# setup.py

setup(
    name="dagster_analytics",
    ...,
+   package_data={"dagster_analytics": ["dbt-project/**/*"]},
)
```

So, if your Dagster folder structure were the following, `dbt-project` will be included as part of the build. Note that we can only include `package_data` [if the package is within the directory of `setup.py`](https://stackoverflow.com/questions/71409139/building-a-package-referencing-a-package-in-a-parent-directory-in-setup-py)

```diff
dbt_repo
└── dagster_analytics
    ├── dagster_analytics
    │   ├── __init__.py
    │   ├── assets.py
    │   ├── constants.py
    │   ├── definitions.py
    │   └── schedules.py
+   ├── dbt-project
    ├── pyproject.toml
    ├── setup.py
└── analyses
└── macros
└── models
└── seeds
└── snapshots
└── target
└── tests
└── dagster_cloud.yaml
└── dbt_project.yml
└── profiles.yml
└── pyproject.toml
└── README.md
```

So, in step (2), we create this `dbt-project` subdirectory in our CI/CD step. We can use `rsync` to copy the `dbt_repo` to `dbt-project`. We exclude `dagster_analytics` in the copy step so we don't have a recursive loop. 

You can see how we do this in our [scaffolded GitHub action](https://github.com/dagster-io/dagster-cloud-action/blob/main/github/serverless/dbt/deploy.yml#L45-L61):

- We run `rsync` to copy the top-level dbt project to the `dbt-project` subdirectory
- We run `dbt deps` and `dbt parse` as part of our guidance for [building the dbt manifest in the CI/CD system](https://docs.dagster.io/integrations/dbt/reference#deploying-a-dbt-project-from-a-monorepo).
- We remove the [`partial_parse.msgpack` file](https://docs.getdbt.com/reference/parsing) to avoid any caching errors when deploying our dbt project.

---

In step (3), you must add the environment variables referenced in your `profiles.yml` to your CI/CD system. This way, dbt can access your warehouse to build your dbt manifest as part of your project.

Furthermore, you should [add the same environment variables to Dagster Cloud](https://docs.dagster.io/dagster-cloud/managing-deployments/environment-variables-and-secrets) so that your code can execute when it is invoked.

If you are using [`dbt-bigquery`](https://docs.getdbt.com/docs/core/connect-data-platform/bigquery-setup), we recommend that you specify a "Service Account JSON", instead of a "Service Account File".

```diff
my-bigquery-db:
  target: dev
  outputs:
    dev:
      type: bigquery
      method: service-account
      project: GCP_PROJECT_ID
      dataset: DBT_DATASET_NAME
      threads: 4 # Must be a value of 1 or greater
-     keyfile: /PATH/TO/BIGQUERY/keyfile.json
+     # Use `keyfile_json` instead. These fields come from the service account json keyfile.
+     #    
+     # These environment variables should be provided in your CI/CD system and Dagster Cloud.
+     keyfile_json:
+       type: xxx
+       project_id: "{{ env_var('MY_DEV_BIGQUERY_PROJECT_ID') }}"
+       private_key_id: "{{ env_var('MY_DEV_BIGQUERY_PRIVATE_KEY_ID') }}"
+       private_key: "{{ env_var('MY_DEV_BIGQUERY_PRIVATE_KEY') }}"
+       client_email: "{{ env_var('MY_DEV_BIGQUERY_CLIENT_EMAIL') }}"
+       client_id: "{{ env_var('MY_DEV_BIGQUERY_CLIENT_ID') }}"
+       auth_uri: "{{ env_var('MY_DEV_BIGQUERY_AUTH_URI') }}"
+       token_uri: "{{ env_var('MY_DEV_BIGQUERY_TOKEN_URI') }}"
+       auth_provider_x509_cert_url: "{{ env_var('MY_DEV_BIGQUERY_AUTH_CERT_URL') }}"
+       client_x509_cert_url: "{{ env_var('MY_DEV_BIGQUERY_CLIENT_CERT_URL') }}"
```

---

Once you do these three steps, you should be able to use the following `dagster_cloud.yaml` file to deploy to Dagster Cloud.

```yaml
locations:
  - location_name: dagster_analytics # the name of the code location
    code_source:
      module_name: "dagster_analytics.definitions"
    build: 
      directory: ./dagster_analytics
```

**Reply by rob-teeuwen** [reply]
@rexledesma thanks for your answer! I added the steps as per your instructions, but the build is still failing with the same error: 

`/venvs/79f23aa32105/lib/python3.8/site-packages/working_directory does not contain a dbt_project.yml file. Please specify a valid path to a dbt project`

I suppose I can fix this by changing my Dagster `constants.py` code, because it still points to the root directory: 

```
dbt_project_dir = Path(__file__).joinpath("..", "..", "..").resolve()
```

However, if I do that, I won't be able to run my project locally anymore, because locally I don't have a CI/CD step that copies the entire dbt project. How is that solved in the scaffolded dbt/dagster-cloud integration? 

Thanks

**Reply by rob-teeuwen** [reply]
Would it make more sense to have the dbt project inside the dagster project, instead of the other way around? It seems that would solve the problem without having to add all this CI/CD stuff. The only reason I have this project structure is because it was provided like this by the dagster-dbt scaffold.

**Reply by rexledesma** [reply]
> I suppose I can fix this by changing my Dagster `constants.py` code, because it still points to the root directory:
> 
> ```
> dbt_project_dir = Path(__file__).joinpath("..", "..", "..").resolve()
> ```
> 
> However, if I do that, I won't be able to run my project locally anymore, because locally I don't have a CI/CD step that copies the entire dbt project. How is that solved in the scaffolded dbt/dagster-cloud integration?

You can use the presence of the [`DAGSTER_CLOUD_DEPLOYMENT_NAME` environment variable](https://docs.dagster.io/dagster-cloud/managing-deployments/environment-variables-and-secrets#all-deployment-variables) to determine whether you should load the manifest from the `dbt-project` subdirectory, or from the top level dbt project in local development.

> Would it make more sense to have the dbt project inside the dagster project, instead of the other way around?

If this is an option for you, then yes that's another solution! However, we don't do this in our onboarding step because creating an automated PR that inverts the directory structure may be a bit confusing for new Dagster users. Introducing an isolated subdirectory to adopt Dagster was a less invasive choice.

**Reply by rob-teeuwen** [reply]
This worked, thank you so much!
